src/wayland-eglsurface.c

```diff
@@ -2564,26 +2523,41 @@ static EGLBoolean wlEglDestroySurface(EGLDisplay dpy, EGLSurface eglSurface)
         surface->wlEventQueue = NULL;
     }
 
-    for (i = 0; i < surface->ctx.numStreamImages; i++) {
-        image = surface->ctx.streamImages[i];
-        pthread_mutex_lock(&image->mutex);
+    if (surface->wlBufferEventQueue) {
+        WlEglStreamImage *image;
+        WlEglStreamImage *nextImage;
 
+        pthread_mutex_lock(&surface->ctx.streamImagesMutex);
         /*
          * Destroy any attached buffers to ensure no further buffer release
          * events are delivered after the buffer release queue and thread are
          * torn down.
+         *
+         * Do not destroy the WlEglStreamImages here because they may be
+         * in-flight and as soon as we unlock streamImagesMutex, the buffer
+         * release thread may access them.
          */
-        if (image->buffer) {
-            assert(image->attached);
-            wl_buffer_destroy(image->buffer);
-            image->buffer = NULL;
-            image->attached = EGL_FALSE;
+        wl_list_for_each(image, &surface->ctx.streamImages, link) {
+            if (image->buffer) {
+                assert(image->attached);
+                wl_buffer_destroy(image->buffer);
+                image->buffer = NULL;
+                image->attached = EGL_FALSE;
+            }
         }
+        pthread_mutex_unlock(&surface->ctx.streamImagesMutex);
 
-        pthread_mutex_unlock(&image->mutex);
-    }
+        finish_wl_buffer_release_thread(surface);
 
-    finish_wl_buffer_release_thread(surface);
+        /*
+         * If there are remaining images in the streamImages list, destroy them
+         * here safely since the buffer release thread is no longer running.
+         */
+        wl_list_for_each_safe(image, nextImage, &surface->ctx.streamImages, link) {
+            destroy_stream_image(display, surface, image);
+        }
```

Comment:
I'm a little confused about the flow here. Shouldn't we only do the attached buffer cancelling and `finish_wl_buffer_release_thread` above when the buffer release thread is in use, but always destroy all the stream images here? Won't this leak them on explicit sync? Or do we know for sure that they will all have been destroyed by this point on explicit sync?

If we know they should all be destroyed it seems like we should mention that in the comments about why this isn't needed on explicit sync?

Reply:
destroy_surface_context() call at the beginning of this function should take care of the stream image destruction when explicit sync is in use. I will add a comment mentioning it.